### PR TITLE
[6X only]Prevent REINDEX TABLE partitioned table statement in transaction block

### PR DESCRIPTION
**Pull request by JunfengYang**
As different from upstream, we support REINDEX TABLE on partitioned
table which will expand the partitioned table and start new transaction
to do the reindex for each table. This has side effects that cannot be
rolled back if it gets called in PL language and may crash, so we need
to make sure it must not run inside a transaction block.
This fix is related to the issue https://github.com/greenplum-db/gpdb/issues/11948.

Tests changes are because the `REINDEX TABLE` on the partitioned table is not
allowed to run in the transaction block now. The behavior is same with upstream,
and the old tests are not suitable, so just update the tests.
Also remove tests which try to test the reindex in transaction block
with sepcific isolation level.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/commands/indexcmds.c [hunk 1]
```diff
@@ -2424,6 +2424,45 @@ ReindexTable(ReindexStmt *stmt)
...
+
+		result = reindex_relation(heapOid,
+								  REINDEX_REL_PROCESS_TOAST |
+									REINDEX_REL_CHECK_CONSTRAINTS);
+		if (!result)
+			ereport(NOTICE,
+					(errmsg("table \"%s\" has no indexes to reindex",
+							stmt->relation->relname)));
+
+		if (result && Gp_role == GP_ROLE_DISPATCH)
```
[the remark this answers is not in the document]
Comment (JunfengYang) [reply src/backend/commands/indexcmds.c:2439]: Yes, you are right.

**Comment by JunfengYang** [comment]
Yes, the removed tests are all test with the below steps
```
1: BEGIN;
1: SET TRANSACTION ISOLATION LEVEL <tested_level>;
1: REINDEX ...
```
This doesn't make any sense right now.

**Review by asimrp — APPROVED** [review]

**Comment by asimrp** [comment]
Please take a look at PR pipeline failure.

File: src/backend/commands/indexcmds.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (JunfengYang) [reply src/backend/commands/indexcmds.c:2439]: We still need check `Gp_role == GP_ROLE_DISPATCH` since the reindex could be executed in utility mode.

**Comment by hidva** [comment]
`reindextable_while_reindex_idx_heap_part_btree` is always failed in my case. 

```
 3: select distinct count(distinct relfilenode), relname from old_relfilenodes group by dbid, relname;
  count | relname
 -------+-----------------------------------------------------
  1     | reindex_crtab_part_heap_btree_1_prt_p_one_id_idx
  1     | reindex_crtab_part_heap_btree_id_idx
  2     | reindex_crtab_part_heap_btree_1_prt_de_fault_id_idx
-(3 rows)
+ 2     | reindex_crtab_part_heap_btree_id_idx
+(4 rows)
 3: COMMIT;
```

The reason is that `REINDEX TABLE reindex_crtab_part_heap_btree` will first reindex reindex_crtab_part_heap_btree after session 1 releases the lock.

```
(gdb) f 10
#10 0x0000000000741110 in ReindexRelationList (relids=0x2a64c48) at indexcmds.c:2274
2274                if (!reindex_relation(relid,
(gdb) p *relids
$8 = {type = T_OidList, length = 3, head = 0x2a64bc8, tail = 0x2a64ca8}
(gdb) p *(ListCell*)0x2a64bc8
// 22715::regclass is reindex_crtab_part_heap_btree
$9 = {data = {ptr_value = 0x3b433d59000058bb, int_value = 22715, oid_value = 22715}, next = 0x2a64c80}
(gdb) p *(ListCell*)0x2a64c80
// 22730::regclass is reindex_crtab_part_heap_btree_1_prt_p_one
$10 = {data = {ptr_value = 0x7f7f7f7f000058ca, int_value = 22730, oid_value = 22730}, next = 0x2a64ca8}
(gdb) p *(ListCell*)0x2a64ca8
// 22722::regclass is reindex_crtab_part_heap_btree_1_prt_de_fault
$11 = {data = {ptr_value = 0x7f7f7f7f000058c2, int_value = 22722, oid_value = 22722}, next = 0x0}
```

**Comment by JunfengYang** [comment]
@hidva Thanks for reporting. The reason should be the lock acquire race condition. I modify the test to make it more stable.

**Comment by ashwinstar** [comment]
we had so many tests before and none exposed the problem, so I am confused a little....
- does the problem only happen if the transaction rolls back?
- what exactly happens when the problem hits crash/error (the github issue says crash but that's only for PL/SQL function not generic transaction block)?

File: src/backend/commands/indexcmds.c [hunk 2]
```diff
@@ -2424,6 +2424,45 @@ ReindexTable(ReindexStmt *stmt)
 	relids = list_concat_unique_oid(relids, prels);
 	MemoryContextSwitchTo(oldcontext);
 
+	if (list_length(relids) == 1)
```
Comment (ashwinstar) [inline src/backend/commands/indexcmds.c:2427]: Missing `MemoryContextDelete(private_context)` for this branch

**Comment by ashwinstar** [comment]
Given the way current code is even for single `REINDEX TABLE` start transaction is done, so I wished to see what will go wrong...
```sql
postgres=# create table foo(a int, b int);
CREATE TABLE
postgres=# create index on foo(b);
CREATE INDEX

postgres=# begin;
BEGIN
postgres=# insert into foo values(1,1);
INSERT 0 1
postgres=# reindex table foo;
REINDEX
postgres=# select relfilenode from pg_class where relname = 'foo_b_idx';
 relfilenode 
-------------
       32775
(1 row)

postgres=# abort;
ROLLBACK

postgres=# select relfilenode from pg_class where relname = 'foo_b_idx';
 relfilenode 
-------------
       32773
(1 row)
```
I didn't encounter any issue. I even tried similarly partition table and still didn't hit any issue.
What am I missing to exhibit the issue?

**Comment by JunfengYang** [comment]
@ashwinstar Have you check the issue https://github.com/greenplum-db/gpdb/issues/11948? We don't have test covered for this case.

**Comment by ashwinstar** [comment]
> @ashwinstar Have you check the issue #11948? We don't have test covered for this case.

yes, I did. I understand the problem. Still curious and wish if the problem is exposed more generically. Hence, I asked earlier what exactly happens when the problem hits crash/error (the github issue says crash but that's only for PL/SQL function not generic transaction block)?

**Comment by JunfengYang** [comment]
@ashwinstar Please check the edited comments above, the message gets sent by accident.

**Comment by JunfengYang** [comment]
@ashwinstar Please check the edited comments above, the message gets sent by accident.

**Comment by JunfengYang** [comment]
@ashwinstar Please also help me review the master version of the fix, which backport the commit mentioned above from upstream.
https://github.com/greenplum-db/gpdb/pull/12003

File: src/backend/commands/indexcmds.c (hunk 2 above)
Comment (JunfengYang) [reply src/backend/commands/indexcmds.c:2427]: Nice catch!

**Comment by hidva** [comment]
> > @ashwinstar Have you check the issue #11948? We don't have test covered for this case.
> 
> yes, I did. I understand the problem. Still curious and wish if the problem is exposed more generically. Hence, I asked earlier what exactly happens when the problem hits crash/error (the github issue says crash but that's only for PL/SQL function not generic transaction block)?

CommitTransactionCommand() will do no-op when executed in a transaction block:

```sql
begin;
-- CommitTransactionCommand() invoked here is no-op.
reindex table partition_table;
commit;
```

And CommitTransactionCommand() executed in an auto-commit context will really call CommitTransaction().

```sql
-- somefunc() is defined in issue#11948
-- CommitTransactionCommand() invoked by `reindex table` is executed in auto-commit context.
-- So CommitTransaction() is called during the execution of PLSQL.
-- So the crash happened because of the reason mentioned by https://github.com/greenplum-db/gpdb/issues/11948#issuecomment-837704265.
select somefunc();
```

**Comment by ashwinstar** [comment]
> > > @ashwinstar Have you check the issue #11948? We don't have test covered for this case.
> > 
> > 
> > yes, I did. I understand the problem. Still curious and wish if the problem is exposed more generically. Hence, I asked earlier what exactly happens when the problem hits crash/error (the github issue says crash but that's only for PL/SQL function not generic transaction block)?
> 
> CommitTransactionCommand() will do no-op when executed in a transaction block:
> 
> ```sql
> begin;
> -- CommitTransactionCommand() invoked here is no-op.
> reindex table partition_table;
> commit;
> ```
> 
> And CommitTransactionCommand() executed in an auto-commit context will really call CommitTransaction().
> 
> ```sql
> -- somefunc() is defined in issue#11948
> -- CommitTransactionCommand() invoked by `reindex table` is executed in auto-commit context.
> -- So CommitTransaction() is called during the execution of PLSQL.
> -- So the crash happened because of the reason mentioned by https://github.com/greenplum-db/gpdb/issues/11948#issuecomment-837704265.
> select somefunc();
> ```

So, problem is only if called in function call. Though `PreventTransactionChain()` prevents us from executing the same in normal transaction block as well. I don't know there are use-cases where some thing is allowed in normal transaction block but only prevented in function calls. For master branch the behavior change seems perfectly fine.
I am just trying to see if possibility exist to avoid changing the behavior on STABLE branch, specially for something which is not broken. REINDEX TABLE anyways doesn't work and crashes if called from UDF so there are for sure no users of it. Though there might be REINDEX TABLE calls in normal transaction blocks possibly which will need unnecessary rewrites after this. Just a thought. Comments?

**Comment by JunfengYang** [comment]
I'm ok with both decisions although I prefer to correct the wrong behavior.
@asimrp Any thoughts?

**Comment by asimrp** [comment]
Ashwin's concern that existing workloads will get severely impacted, is valid.  Once existing deployments are upgraded to the 6X version that contains this fix, workloads having `begin; reindex <partition table>; ... end;` will stop working.

The most significant drawback of allowing such workloads is locks on the entire partition hierarchy will be held until end of the transaction.  If things have worked without complaints for so long, avoiding the disruption in workload is desirable.

It will be a good idea to emit a deprecation warning for such workloads, indicating that newer version of Greenplum will no longer allow `reindex` with an explicit transaction.  What do you think @wenwangBJ, @wuj21? 

There is no second thought on fixing the crash, which happens when reindex is executed from a UDF.

**Comment by asimrp** [comment]
Note that workloads of the form `{{begin; ... reindex <non-partition table>; ... end;}}` are NOT impacted by this patch.

**Comment by ashwinstar** [comment]
So, what's the consensus for this PR, Do folks feel we should keep it in current form or reduce the scope to only ERROR for UDF case?

**Comment by JunfengYang** [comment]
Any more feedbacks? @baobao0206 @adam8157 @kainwen

**Comment by darthunix** [comment]
I wonder why do GPDB and [PG](https://github.com/postgres/postgres/commit/a6642b3ae060976b42830b7dc8f29ec190ab05e4) use the same per-transaction approach to reindex partitioned tables? I have an explanation in PG commit, but didn't get it:

> This shares some logic with schema and database REINDEX as each
partition gets processed in its own transaction after building a list of
relations to work on.  This choice has the advantage to minimize the
number of invalid indexes to one partition with REINDEX CONCURRENTLY in
the event a cancellation or failure in-flight, as the only indexes
handled at once in a single REINDEX CONCURRENTLY loop are the ones from
the partition being working on.

On transaction rollback we always trim and unlink index files. Is it such a performance expensive operation that we don't want to execute it more than the number of indexes a single partition has? Cause if we reindex all partitions in a single transaction we don't need to forbid reindex in plpgsql/subtransactions, right?

**Comment by ashwinstar** [comment]
Closing this in favor of https://github.com/greenplum-db/gpdb/pull/14081

**Pull request closed without merging**